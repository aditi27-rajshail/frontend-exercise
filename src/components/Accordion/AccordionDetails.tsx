/**
 * AccordionDetails component for displaying detailed content in an accordion-style container.
 * @param {AccordionDetailsData} accordionDetails - The data for the accordion details.
 * @param {Function} onToggle - The function to handle the toggle action.
 * @param {boolean} isChecked - A flag to indicate whether the item is checked.
 * @param {boolean} disabled - A flag to indicate whether the accordion details are disabled.
 * @returns {React.Element} - The rendered AccordionDetails component.
 */

import Button from "../Button";
import { checkCircle, circleDashed } from "../../assets";
import Alert from "../Alert";
import { AccordionDetailsData } from "../../types";
import { actionBtnHandlers } from "../../logic/actionBtnHandlers";

interface AccordionDetailsProps {
  accordionDetails: AccordionDetailsData;
  onToggle: () => void;
  isChecked: boolean;
  disabled: boolean;
}

const AccordionDetails = ({
  accordionDetails,
  onToggle,
  isChecked,
  disabled,
}: AccordionDetailsProps) => {
  const callActionHandler = (actionKey: keyof typeof actionBtnHandlers) => {
    const handler = actionBtnHandlers[actionKey];
    if (handler) {
      handler();
    } else {
      console.log(`No action handler for key: ${actionKey}`);
    }
  };
  return (
    <div className="border border-solid border-neutral500 mt-4 p-4 rounded-lg">
      <div className="md:flex lg:gap-[77px] md:gap-[38px]">
        <div className="flex items-start">
          <img
            src={isChecked ? checkCircle : circleDashed}
            alt={isChecked ? "check circle" : "circle dashed"}
            width={24}
            height={24}
            className={` mr-[8px] ${!disabled && "cursor-pointer easeout"}`}
            onClick={!disabled ? onToggle : undefined}
          />
          {/* Content */}
          <div>
            <div className={`${disabled && "disabled"}`}>
              <h4 className="accordion-detail-title mb-2">
                {accordionDetails.title}
              </h4>
              {accordionDetails?.tag && (
                <span className="chip text-info600 h-[20px] py-0 px-2">
                  {accordionDetails.tag}
                </span>
              )}

              <p className="accordion-detail-sub-text mt-2">
                {accordionDetails.subText}
              </p>
            </div>
            {accordionDetails?.info && (
              <Alert text={accordionDetails.info} severity="info" />
            )}
          </div>
        </div>
        <div className="flex flex-col items-center justify-center mt-3 md:mt-0">
          <Button
            buttonText={accordionDetails.buttonText}
            disabled={disabled}
            onClick={() => callActionHandler(accordionDetails.btnAction)}
          />
        </div>
      </div>
    </div>
  );
};

export default AccordionDetails;
