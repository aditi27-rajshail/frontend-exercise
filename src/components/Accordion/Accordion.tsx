/**
 * Accordion component for displaying a section with expandable/collapsible content.
 * @param {MainData} accordionData - The data for the accordion section.
 * @returns {React.Element} - The rendered Accordion component.
 */

import { useState, useCallback } from "react";
import { caretUp } from "../../assets";
import ProgressBar from "../ProgressBar";
import AccordionDetails from "./AccordionDetails";
import { AccordionDetailsData, MainData } from "../../types";

interface AccordionProps {
  accordionData: MainData;
}
const Accordion = ({ accordionData }: AccordionProps) => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const [checkedState, setCheckedState] = useState<boolean[]>(
    new Array(accordionData.accordionDetails.length).fill(false)
  );

  const toggleChecked = useCallback((index: number) => {
    setCheckedState((prevState) => {
      const updatedCheckedState = [...prevState];
      updatedCheckedState[index] = !updatedCheckedState[index];
      return updatedCheckedState;
    });
  }, []);

  const stepsCompleted = checkedState.filter(Boolean).length;
  return (
    <div className="bg-white100 rounded-lg card flex flex-col items-start mb-4 w-full xl:w-[1011px] p-4 ">
      {/* Header */}
      <div className="flex justify-between w-full items-start xl:items-center">
        <div className="flex items-center">
          <img
            src={accordionData.img}
            alt={accordionData.title}
            width={26}
            height={26}
            className="mr-2"
          />
          <h3 className="card-title">{accordionData.title}</h3>
        </div>
        <div className="flex items-center space-x-2">
          <div className="hidden sm:block">
            <ProgressBar
              step={stepsCompleted}
              totalSteps={accordionData.accordionDetails.length}
            />
          </div>
          <button
            onClick={() => setIsOpen(!isOpen)}
            className="transform transition-transform w-[16px] h-[16px] flex items-center justify-center"
            aria-label={isOpen ? "Collapse section" : "Expand section"}
            aria-expanded={isOpen}>
            <img
              src={caretUp}
              width={16}
              height={16}
              alt="Toggle accordion"
              className={`transition-transform duration-500 ${
                isOpen ? "rotate-180" : "rotate-0"
              }`}
            />
          </button>
        </div>
      </div>
      <p className="cardSubText my-4 sm:my-4">{accordionData.subText}</p>
      <div className="w-full mb-2 sm:hidden">
        <ProgressBar
          step={stepsCompleted}
          totalSteps={accordionData.accordionDetails.length}
        />
      </div>

      {/* Accordion Content */}

      {isOpen &&
        accordionData.accordionDetails.map(
          (accordionDetails: AccordionDetailsData, index: number) => {
            return (
              <AccordionDetails
                key={accordionDetails.title}
                accordionDetails={accordionDetails}
                onToggle={() => toggleChecked(index)}
                isChecked={checkedState[index]}
                disabled={accordionData.isDisabled}
              />
            );
          }
        )}
    </div>
  );
};

export default Accordion;
