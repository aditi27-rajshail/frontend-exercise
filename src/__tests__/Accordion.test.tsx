import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, beforeEach } from "vitest";
import Accordion from "../components/Accordion/Accordion";
import { mainData } from "../constants";

describe("Accordion Component", () => {
  beforeEach(() => {
    render(<Accordion accordionData={mainData[0]} />);
  });

  it("renders the accordion title", () => {
    expect(screen.getByText(mainData[0].title).textContent).toBe(
      mainData[0].title
    );
  });

  it("renders the accordion subText", () => {
    expect(screen.getByText(mainData[0].subText).textContent).toBe(
      mainData[0].subText
    );
  });

  it("toggles accordion content on click", () => {
    // Gets the toggele button
    const toggleButton = screen.getByRole("button", {
      name: /expand section/i,
    });
    // Click to open
    fireEvent.click(toggleButton);
    expect(toggleButton.getAttribute("aria-expanded")).toBe("true");
    // Click to close
    fireEvent.click(toggleButton);
    expect(toggleButton.getAttribute("aria-expanded")).toBe("false");
  });

  it("matches the snapshot when closed", () => {
    const { container } = render(<Accordion accordionData={mainData[0]} />);
    expect(container).toMatchSnapshot();
  });

  it("matches the snapshot when opened", () => {
    const { container } = render(<Accordion accordionData={mainData[0]} />);
    const allToggleButtons = screen.getAllByRole("button", {
      name: /expand section/i,
    });
    // Open the first accordion
    fireEvent.click(allToggleButtons[0]);
    expect(container).toMatchSnapshot();
  });

  it("updates the ProgressBar when a check icon is checked", () => {
    // open accordion
    const toggleButton = screen.getByRole("button", {
      name: /expand section/i,
    });
    fireEvent.click(toggleButton);
    expect(toggleButton.getAttribute("aria-expanded")).toBe("true");

    // get check icon and click
    const checkboxImage = screen.getAllByRole("img", {
      name: /circle dashed/i,
    });
    fireEvent.click(checkboxImage[0]);

    const progressBarTextElements = screen.getAllByTestId("progress-text");
    progressBarTextElements.forEach((element) => {
      // remove whitespace
      expect(element.textContent).toMatch(/1\s*\/\s*2\s*ステップ/);
    });
  });
});
