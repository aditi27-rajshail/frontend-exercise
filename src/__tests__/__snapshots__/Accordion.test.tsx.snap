// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion Component > matches the snapshot when closed 1`] = `
<div>
  <div
    class="bg-white100 rounded-lg card flex flex-col items-start mb-4 w-full xl:w-[1011px] p-4 "
  >
    <div
      class="flex justify-between w-full items-start xl:items-center"
    >
      <div
        class="flex items-center"
      >
        <img
          alt="取引の情報を入力しましょう"
          class="mr-2"
          height="26"
          src="/src/assets/user-circle-gear-duotone.svg"
          width="26"
        />
        <h3
          class="card-title"
        >
          取引の情報を入力しましょう
        </h3>
      </div>
      <div
        class="flex items-center space-x-2"
      >
        <div
          class="hidden sm:block"
        >
          <div
            class="flex items-center"
          >
            <div
              class="h-[6px] w-[60px] bg-neutral500 rounded overflow-hidden mr-2"
            >
              <div
                class="bg-green500  transition-all duration-300 h-full"
                style="width: 0%;"
              />
            </div>
            <span
              class="progress-text"
              data-testid="progress-text"
            >
              0
              /
              2
              ステップ
            </span>
          </div>
        </div>
        <button
          aria-expanded="false"
          aria-label="Expand section"
          class="transform transition-transform w-[16px] h-[16px] flex items-center justify-center"
        >
          <img
            alt="Toggle accordion"
            class="transition-transform duration-500 rotate-0"
            height="16"
            src="/src/assets/caret-up-bold.svg"
            width="16"
          />
        </button>
      </div>
    </div>
    <p
      class="cardSubText my-4 sm:my-4"
    >
      仮想通貨の損益計算には過去すべての取引履歴が必要になります。
    </p>
    <div
      class="w-full mb-2 sm:hidden"
    >
      <div
        class="flex items-center"
      >
        <div
          class="h-[6px] w-[60px] bg-neutral500 rounded overflow-hidden mr-2"
        >
          <div
            class="bg-green500  transition-all duration-300 h-full"
            style="width: 0%;"
          />
        </div>
        <span
          class="progress-text"
          data-testid="progress-text"
        >
          0
          /
          2
          ステップ
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion Component > matches the snapshot when opened 1`] = `
<div>
  <div
    class="bg-white100 rounded-lg card flex flex-col items-start mb-4 w-full xl:w-[1011px] p-4 "
  >
    <div
      class="flex justify-between w-full items-start xl:items-center"
    >
      <div
        class="flex items-center"
      >
        <img
          alt="取引の情報を入力しましょう"
          class="mr-2"
          height="26"
          src="/src/assets/user-circle-gear-duotone.svg"
          width="26"
        />
        <h3
          class="card-title"
        >
          取引の情報を入力しましょう
        </h3>
      </div>
      <div
        class="flex items-center space-x-2"
      >
        <div
          class="hidden sm:block"
        >
          <div
            class="flex items-center"
          >
            <div
              class="h-[6px] w-[60px] bg-neutral500 rounded overflow-hidden mr-2"
            >
              <div
                class="bg-green500  transition-all duration-300 h-full"
                style="width: 0%;"
              />
            </div>
            <span
              class="progress-text"
              data-testid="progress-text"
            >
              0
              /
              2
              ステップ
            </span>
          </div>
        </div>
        <button
          aria-expanded="false"
          aria-label="Expand section"
          class="transform transition-transform w-[16px] h-[16px] flex items-center justify-center"
        >
          <img
            alt="Toggle accordion"
            class="transition-transform duration-500 rotate-0"
            height="16"
            src="/src/assets/caret-up-bold.svg"
            width="16"
          />
        </button>
      </div>
    </div>
    <p
      class="cardSubText my-4 sm:my-4"
    >
      仮想通貨の損益計算には過去すべての取引履歴が必要になります。
    </p>
    <div
      class="w-full mb-2 sm:hidden"
    >
      <div
        class="flex items-center"
      >
        <div
          class="h-[6px] w-[60px] bg-neutral500 rounded overflow-hidden mr-2"
        >
          <div
            class="bg-green500  transition-all duration-300 h-full"
            style="width: 0%;"
          />
        </div>
        <span
          class="progress-text"
          data-testid="progress-text"
        >
          0
          /
          2
          ステップ
        </span>
      </div>
    </div>
  </div>
</div>
`;
